pgserver/subscription_handler.go

```diff
@@ -0,0 +1,216 @@
+package pgserver
+
+import (
+	"context"
+	stdsql "database/sql"
+	"fmt"
+	"github.com/apecloud/myduckserver/adapter"
+	"github.com/apecloud/myduckserver/pgserver/logrepl"
+	"github.com/jackc/pglogrepl"
+	"regexp"
+)
+
+// This file implements the logic for handling CREATE SUBSCRIPTION SQL statements.
+// Example usage of CREATE SUBSCRIPTION SQL:
+//
+// CREATE SUBSCRIPTION mysub
+// CONNECTION 'dbname= host=127.0.0.1 port=15432 user=postgres password=root'
+// PUBLICATION mypub;
+//
+// The statement creates a subscription named 'mysub' that connects to a PostgreSQL
+// database and subscribes to changes published under the 'mypub' publication.
+
+type SubscriptionConfig struct {
+	SubscriptionName string
+	PublicationName  string
+	DBName           string
+	Host             string
+	Port             string
+	User             string
+	Password         string
+	LSN              pglogrepl.LSN
+}
+
+var lsnQueryIndex = 5
+
+var subscriptionRegex = regexp.MustCompile(`CREATE SUBSCRIPTION\s+(\w+)\s+CONNECTION\s+'([^']+)'\s+PUBLICATION\s+(\w+);`)
+var connectionRegex = regexp.MustCompile(`(\b\w+)=([\w\.\d]*)`)
+
+// ToConnectionInfo Format SubscriptionConfig into a ConnectionInfo
+func (config *SubscriptionConfig) ToConnectionInfo() string {
+	return fmt.Sprintf("dbname=%s user=%s password=%s host=%s port=%s",
+		config.DBName, config.User, config.Password, config.Host, config.Port)
+}
+
+// ToDNS Format SubscriptionConfig into a DNS
+func (config *SubscriptionConfig) ToDNS() string {
+	return fmt.Sprintf("postgres://%s:%s@%s:%s/%s",
+		config.User, config.Password, config.Host, config.Port, config.DBName)
+}
+
+func (config *SubscriptionConfig) ToDuckDBQuery() []string {
+	return []string{
+		"INSTALL postgres_scanner;",
+		"LOAD postgres_scanner;",
+		fmt.Sprintf("ATTACH '%s' AS pg_postgres (TYPE POSTGRES);", config.ToConnectionInfo()),
+		"BEGIN;",
+		"COPY FROM DATABASE pg_postgres TO mysql;",
+		"SELECT * FROM postgres_query('pg_postgres', 'SELECT pg_current_wal_lsn()');",
+		"COMMIT;",
```

Comment:
Should it be detached when the copy is done?

Reply:
fixed